### PR TITLE
refactor extension types and begin remote (runtime) extension support

### PR DESCRIPTION
**Pull request by kckst8**
I refactored the types to streamline the 2 types of extensions (local/remote(runtime)). If an extension provides a url for a remote extension, we are expecting it to be a module with a default export that is a function that begins the execution of its code. The import/execution will occur after its first activation event. There are certainly some cons and fragility to this approach, and even if accepted it needs some optimization and cleanup, which is why I am looking for feedback. 

This PR would be required to support dynamic remote imports with Webpack and commonjs: https://github.com/imodeljs/create-react-app/pull/69

This shows an example of a local and remote extension being provided to the Viewer with this approach: https://github.com/iTwin/viewer/pull/150

### CODE REVIEW

File: core/frontend/src/extension/Extension.ts
Comment (johnnyd710) [inline core/frontend/src/extension/Extension.ts:76]: Would the use of @aruniverse's XOR type be useful here instead of all these optional parameters?
Something like `type ExtensionProvider = XOR<LocalExtensionProvider, RemoteExtensionProvider>` since it does not make sense to have both properties

Comment (aruniverse) [reply core/frontend/src/extension/Extension.ts:76]: I was thinking of maybe moving the xor type into core\bentley\src\UtilityTypes.ts

Comment (kabentley) [reply core/frontend/src/extension/Extension.ts:76]: where is it now?

Comment (aruniverse) [reply core/frontend/src/extension/Extension.ts:76]: Sitting in a branch for a [pr](https://github.com/iTwin/viewer/pull/149) at the moment, https://github.com/iTwin/viewer/blob/types-next/packages/modules/viewer-react/src/types.ts#L30-L33

Comment (kckst8) [reply core/frontend/src/extension/Extension.ts:76]: One of the goals of this PR is to blur the lines between local and remote extensions. For example, an extension should be able to provide its javascript at build time but still provide its manifest remotely if it so chooses. I don't believe that we should care to force a method to provide the bundle based on how a user provides the manifest and vice versa.

That said, it does make sense to make the 2 methods to provide the js and the 2 methods to provide the manifest OR. I can update those and swap for the new type if/when added

Comment (kckst8) [reply core/frontend/src/extension/Extension.ts]: changed to a type with xor properties

File: core/frontend/src/extension/Extension.ts
Comment (pmconne) [reply core/frontend/src/extension/Extension.ts:43]: IMO ActivationEvent should be a string union, not an enum.

Comment (kckst8) [reply core/frontend/src/extension/Extension.ts:43]: Perhaps back to string literals is the better option. I'll update

Comment (aruniverse) [reply core/frontend/src/extension/Extension.ts:43]: That change was made in this commit, [`d6fb1b4` (#3146)](https://github.com/iTwin/itwinjs-core/pull/3146/commits/d6fb1b455ff3d9a39b2f6ce7fd8352c0605838b3)

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (kckst8) [reply core/frontend/src/extension/ExtensionAdmin.ts:80]: isn't that an es2020 feature? It would make sense here, but we would need to upgrade our targets (which I think we should anyways)

Comment (aruniverse) [reply core/frontend/src/extension/ExtensionAdmin.ts:80]: Ah good point, im hoping we can switch to es2020 along with the 3.2 release, just need to wrap up that webpack5 pre-reqs

File: core/frontend/src/extension/Extension.ts
Comment (calebmshafer) [reply core/frontend/src/extension/Extension.ts:43]: @aruniverse any reason to not resolve this?

**Comment by NancyMcCallB** [comment]
Looks like the doc generation removed some of the APIs from the doc for appui-react and appui-layout-react. Perhaps npm run build && npm run extract-api in those two directories to correct it? (Rush sometimes fails to do this correctly after a merge, though I don't know why.)

**Comment by dassaf4** [comment]
> Looks like the doc generation removed some of the APIs from the doc for appui-react and appui-layout-react. Perhaps npm run build && npm run extract-api in those two directories to correct it? (Rush sometimes fails to do this correctly after a merge, though I don't know why.)

Recent changes in common/api/core-geometry.api.md are also not showing up in the diff...

**Comment by kckst8** [comment]
> Looks like the doc generation removed some of the APIs from the doc for appui-react and appui-layout-react. Perhaps npm run build && npm run extract-api in those two directories to correct it? (Rush sometimes fails to do this correctly after a merge, though I don't know why.)

nice catch. Not sure what happened there. I re-ran and updated this PR. Can you verify that everything looks ok now?

File: core/frontend/src/extension/Extension.ts
Comment (pmconne) [inline core/frontend/src/extension/Extension.ts:12]: Why is this being exported from core-frontend?

**Comment by pmconne** [comment]
> There are certainly some cons and fragility to this approach

Can you elaborate please?

File: core/frontend/src/extension/Extension.ts
Comment (johnnyd710) [reply core/frontend/src/extension/Extension.ts:12]: Sorry I forgot to change the "imodeljs" from the previous iteration of Extensions (v2.19).
Actually I just discovered the FrontendLoggerCategory enum... should I change the value to "core-frontend.Extensions" and add to the enum?

Comment (pmconne) [reply core/frontend/src/extension/Extension.ts:12]: That sounds right.

File: core/frontend/src/extension/providers/ExtensionServiceClient.ts
Comment (calebmshafer) [inline core/frontend/src/extension/providers/ExtensionServiceClient.ts:1]: Header missing somehow...

File: core/extension/index.d.ts
Comment (calebmshafer) [inline core/extension/index.d.ts:52]: Any reason these moved from where they are on `master`? Did you re-run the generation?

File: core/frontend/src/FrontendLoggerCategory.ts
Comment (calebmshafer) [inline core/frontend/src/FrontendLoggerCategory.ts:42]: Can you make this `alpha`

File: core/extension/index.d.ts
Comment (johnnyd710) [reply core/extension/index.d.ts:52]: It was ran a while back: [this commit](https://github.com/iTwin/itwinjs-core/pull/3347/commits/ea00efe1a9ebaa7233a18cc1d99e449ebab077d2)

Comment (kckst8) [reply core/extension/index.d.ts:52]: probably best to merge in master again and re-run the generation to be safe.

File: core/frontend/src/extension/providers/ExtensionServiceClient.ts
Comment (kckst8) [inline core/frontend/src/extension/providers/ExtensionServiceClient.ts:1]: in hindsight, we should move this into its own repo/package after we release 3.2 since the extension CLI will need it as well once we bring it back. Nothing to do for now, just noting

**Review by kckst8 — COMMENTED** [review]
I'm not able to approve since I'm the original author, but this LGTM once comments are resolved. I tested remote, local, and from the service and all worked as expected. Nice work on this @johnnyd710 !

**Review by pmconne — APPROVED** [review]

File: core/frontend/src/extension/Extension.ts
Comment (calebmshafer) [inline core/frontend/src/extension/Extension.ts:74]: Why do we have the `execute` on both the ExtensionProvider and InstalledExtension? It seems like it'd only be necessary on the `InstalledExtension`.

Edit: I see why there's two now since the `execute` on the ExtensionProvider is really the "download" function that we had previously and is just in charge of getting that entry module but not technically executing it...

File: core/frontend/src/extension/providers/ServiceExtensionProvider.ts
Comment (calebmshafer) [inline core/frontend/src/extension/providers/ServiceExtensionProvider.ts:16]: Let's avoid the term "service" extension here as that's fairly confusing plus it doesn't look to use the `LocalExtensionProps`.

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (calebmshafer) [inline core/frontend/src/extension/ExtensionAdmin.ts:103]: We decided to not use `void` to avoid using the eslint-disable-line and instead keep that warning in there. I've seen this in a few cases, mind switching them back?

File: core/frontend/src/extension/Extension.ts
Comment (calebmshafer) [inline core/frontend/src/extension/Extension.ts:52]: Should `InstalledExtension` be a user facing concept here? It seems to be the internal detail of the ExtensionAdmin and can be moved to that file and not exported.

File: core/frontend/src/extension/providers/ExtensionServiceClient.ts
Comment (calebmshafer) [inline core/frontend/src/extension/providers/ExtensionServiceClient.ts:24]: This should be private.

File: core/frontend/src/extension/Extension.ts
Comment (calebmshafer) [inline core/frontend/src/extension/Extension.ts:83]: nit: Can these be moved next to their respective Providers?

File: core/frontend/src/extension/Extension.ts
Comment (calebmshafer) [inline core/frontend/src/extension/Extension.ts:117]: Can we move this, and the other two interfaces, to the `ServiceExtensionProvider.ts`? I think this type name should probably be used for something else instead of just directly describing the metadata about an Extension from Bentley's Extension Service directly...

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (calebmshafer) [inline core/frontend/src/extension/ExtensionAdmin.ts:69]: same comment here about the `void`

File: core/frontend/src/extension/Extension.ts
Comment (calebmshafer) [inline core/frontend/src/extension/Extension.ts:111]: This should be switched to iTwinId

**Review by calebmshafer — COMMENTED** [review]
If I'm reading this correctly, the main change is that each Extension is going to require the creation of an instance of a Provider class that will be implicitly kept around to know exactly where to get the Extension from. This is instead of the alternative where we previously iterated over each of the Loaders that an Extension could be gathered from which ended up in potentially useless server calls. Please correct me if I'm mis-stating anything there.

It puts a bit more onerous on whoever is adding the Extension in that they need to know exactly where it's coming from but still not let it come from two different places. This may have ramifications on what the API needs to look like when you're wrapping this in an application to allow the local development workflow, and side-loading, in the Viewer. Unless the iTwin Viewer always either loads from one or the other. I need to think this part through some more...

**Comment by calebmshafer** [comment]
> > There are certainly some cons and fragility to this approach
> 
> Can you elaborate please?

@pmconne @kckst8 @johnnyd710 I have the same question. What cons do you see with this approach? I listed some in my other [comment](https://github.com/iTwin/itwinjs-core/pull/3347#pullrequestreview-968490326) but feel like I'm missing something here when reading through the code.

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (kckst8) [reply core/frontend/src/extension/ExtensionAdmin.ts:20]: I think it should be 3 but we need to add a 3rd for the extension service

**Comment by kckst8** [comment]
> If I'm reading this correctly, the main change is that each Extension is going to require the creation of an instance of a Provider class that will be implicitly kept around to know exactly where to get the Extension from. This is instead of the alternative where we previously iterated over each of the Loaders that an Extension could be gathered from which ended up in potentially useless server calls. Please correct me if I'm mis-stating anything there.
> 
> It puts a bit more onerous on whoever is adding the Extension in that they need to know exactly where it's coming from but still not let it come from two different places. This may have ramifications on what the API needs to look like when you're wrapping this in an application to allow the local development workflow, and side-loading, in the Viewer. Unless the iTwin Viewer always either loads from one or the other. I need to think this part through some more...

This is correct. @johnnyd710  proposed this provider architecture and it made sense to me. We've been lumping a group of related but contractually different items into a single somewhat ambiguous item called an "Extension" and then attempting to decipher what each actually is internally, like you mentioned. I believe this adds more confusion for the user and is more fragile for the API, as we are essentially guessing and making assumptions as to what the user wants in the current architecture. While this adds an extra step for consumers, it makes it abundantly clear to them and our API what they are attempting to load. Unless I am missing something, I don't know that there is a case where an application developer would be attempting to load an extension without knowing its source. We just need to make sure that we document each clearly. 

As for the viewer, currently it accepts an array of instances of ExtensionProviders, which could be local, remote, or from the service. We could overload its type and have the user just pass in extensions and a type and let the Viewer create the provider instances, but I'm not so sure that we want to hide the differences from the user. Its very declarative.

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (kckst8) [reply core/frontend/src/extension/ExtensionAdmin.ts:103]: are you suggesting to re-add eslint-disable or just allow the warnings to occur? I don't love the void hack either, but I dont know that we should just ignore the warnings?

File: core/frontend/src/extension/providers/ServiceExtensionProvider.ts
Comment (kckst8) [reply core/frontend/src/extension/providers/ServiceExtensionProvider.ts:16]: @johnnyd710  and I both agree. We spoke about this, more in the context of the type and interface names, and decided to get others opinions in this PR. I could not come up with a meaningful way to express this type of extension. Everything that we came up with was not unique to this case.

**Comment by kckst8** [comment]
> > > There are certainly some cons and fragility to this approach
> > 
> > 
> > Can you elaborate please?
> 
> @pmconne @kckst8 @johnnyd710 I have the same question. What cons do you see with this approach? I listed some in my other [comment](https://github.com/iTwin/itwinjs-core/pull/3347#pullrequestreview-968490326) but feel like I'm missing something here when reading through the code.

Listed most of my thoughts in my comment below. I don't know of anything that makes this more fragile, other than the potential for dynamic import to fail based on build system and module format. However, @johnnyd710  implemented a fallback for that case that I believe is sound. @johnnyd710 was previously working on a solution that involved script tags. Possibly this comment was leftover from that?

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (calebmshafer) [reply core/frontend/src/extension/ExtensionAdmin.ts:103]: re-add the eslint-disable.

**Comment by calebmshafer** [comment]
> > If I'm reading this correctly, the main change is that each Extension is going to require the creation of an instance of a Provider class that will be implicitly kept around to know exactly where to get the Extension from. This is instead of the alternative where we previously iterated over each of the Loaders that an Extension could be gathered from which ended up in potentially useless server calls. Please correct me if I'm mis-stating anything there.
> > It puts a bit more onerous on whoever is adding the Extension in that they need to know exactly where it's coming from but still not let it come from two different places. This may have ramifications on what the API needs to look like when you're wrapping this in an application to allow the local development workflow, and side-loading, in the Viewer. Unless the iTwin Viewer always either loads from one or the other. I need to think this part through some more...
> 
> This is correct. @johnnyd710 proposed this provider architecture and it made sense to me. We've been lumping a group of related but contractually different items into a single somewhat ambiguous item called an "Extension" and then attempting to decipher what each actually is internally, like you mentioned. I believe this adds more confusion for the user and is more fragile for the API, as we are essentially guessing and making assumptions as to what the user wants in the current architecture. While this adds an extra step for consumers, it makes it abundantly clear to them and our API what they are attempting to load. Unless I am missing something, I don't know that there is a case where an application developer would be attempting to load an extension without knowing its source. We just need to make sure that we document each clearly.
> 
> As for the viewer, currently it accepts an array of instances of ExtensionProviders, which could be local, remote, or from the service. We could overload its type and have the user just pass in extensions and a type and let the Viewer create the provider instances, but I'm not so sure that we want to hide the differences from the user. Its very declarative.

@kckst8 okay, that makes sense to me. It took me about 5 reads through the code to understand exactly how of it all worked due to the multiple `execute` methods and how they're invoked :) However, now that I know I get the reasoning for doing it.

In the Viewer are you going to introduce a prop for a host url to be added? That's one of the strange parts about the API, that function is only valid in the case that you're configuring your own Extension Service. I couldn't think of a better way to do it though in order to avoid people trying to load Extensions from just any random url.

**Comment by kckst8** [comment]
> > > If I'm reading this correctly, the main change is that each Extension is going to require the creation of an instance of a Provider class that will be implicitly kept around to know exactly where to get the Extension from. This is instead of the alternative where we previously iterated over each of the Loaders that an Extension could be gathered from which ended up in potentially useless server calls. Please correct me if I'm mis-stating anything there.
> > > It puts a bit more onerous on whoever is adding the Extension in that they need to know exactly where it's coming from but still not let it come from two different places. This may have ramifications on what the API needs to look like when you're wrapping this in an application to allow the local development workflow, and side-loading, in the Viewer. Unless the iTwin Viewer always either loads from one or the other. I need to think this part through some more...
> > 
> > 
> > This is correct. @johnnyd710 proposed this provider architecture and it made sense to me. We've been lumping a group of related but contractually different items into a single somewhat ambiguous item called an "Extension" and then attempting to decipher what each actually is internally, like you mentioned. I believe this adds more confusion for the user and is more fragile for the API, as we are essentially guessing and making assumptions as to what the user wants in the current architecture. While this adds an extra step for consumers, it makes it abundantly clear to them and our API what they are attempting to load. Unless I am missing something, I don't know that there is a case where an application developer would be attempting to load an extension without knowing its source. We just need to make sure that we document each clearly.
> > As for the viewer, currently it accepts an array of instances of ExtensionProviders, which could be local, remote, or from the service. We could overload its type and have the user just pass in extensions and a type and let the Viewer create the provider instances, but I'm not so sure that we want to hide the differences from the user. Its very declarative.
> 
> @kckst8 okay, that makes sense to me. It took me about 5 reads through the code to understand exactly how of it all worked due to the multiple `execute` methods and how they're invoked :) However, now that I know I get the reasoning for doing it.
> 
> In the Viewer are you going to introduce a prop for a host url to be added? That's one of the strange parts about the API, that function is only valid in the case that you're configuring your own Extension Service. I couldn't think of a better way to do it though in order to avoid people trying to load Extensions from just any random url.

@calebmshafer same thoughts and I don't know that I have a great solution yet. Basically I expect that the Viewer will iterate and register the hosts, but we should be explicit that it is being done and allow the user to potentially even disallow, though I'm not sure why they would

File: core/frontend/src/extension/Extension.ts
Comment (johnnyd710) [reply core/frontend/src/extension/Extension.ts:74]: Technically, when execute() on the ExtensionProvider is called, the code is downloaded and executed at the same time. Actually there is no concept of a difference between "download" vs "execute" anymore. I understand the confusion... originally I put the whole Provider inside the InstalledExtension like so `{ provider: ExtensionProvider, manifest: ExtensionManifest }`. But now when ExtensionAdmin wants to execute the extension its more verbose: `extension.provider.execute()` instead of just `extension.execute()`. Thought it was better to get rid of providers at this point since they've already done their job

File: core/frontend/src/extension/providers/ServiceExtensionProvider.ts
Comment (johnnyd710) [reply core/frontend/src/extension/providers/ServiceExtensionProvider.ts:16]: @calebmshafer what do we want to call Extensions that are hosted on the Extension Service?

File: core/frontend/src/extension/Extension.ts
Comment (johnnyd710) [reply core/frontend/src/extension/Extension.ts:52]: Yeah definitely, its just an ExtensionAdmin concept. Will move

File: core/frontend/src/extension/Extension.ts
Comment (johnnyd710) [reply core/frontend/src/extension/Extension.ts:117]: I agree I thought ExtensionProps was confusing as well... seems like everything is a prop! Does ExtensionMetadata sound better?

File: core/frontend/src/extension/Extension.ts
Comment (johnnyd710) [reply core/frontend/src/extension/Extension.ts:111]: I can change it in some things but not in the "ExtensionProps" (or what I proposed to call ExtensionMetadata) until we change the actual service api

File: core/frontend/src/extension/providers/ExtensionLoadScript.ts
Comment (johnnyd710) [inline core/frontend/src/extension/providers/ExtensionLoadScript.ts:17]: @kckst8 @calebmshafer be aware that I had to comment out the original dynamic import. Otherwise, the build will not pass (specifically, the extract-api step) because of a warning "Critical Dependency: the request of a dependency is an expression". I'm still trying to figure out a way around this... I don't think we should just ignore the warning, do we?

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (calebmshafer) [reply core/frontend/src/extension/ExtensionAdmin.ts:20]: I'd just call it the Bentley Extension Service for now. We still have some work to do for it to be in iTwin platform properly and expose it with the correct naming.

File: core/frontend/src/extension/Extension.ts
Comment (calebmshafer) [reply core/frontend/src/extension/Extension.ts:62]: Yea I guess that's where I was getting mixed up, there isn't a difference between the two of them. However, I wonder if we do just keep the `ExtensionProvider` in the `InstalledExtension` to make it explicit. Especially since it's just an internal concept of the `ExtensionAdmin` the added info isn't too bad.

In addition, I think the `ExtensionProvider` would end up not being cleaned up anyway since it would own the execute method that's set as the execute method.

File: core/frontend/src/extension/providers/ServiceExtensionProvider.ts
Comment (calebmshafer) [reply core/frontend/src/extension/providers/ServiceExtensionProvider.ts:29]: I'd just refer to it as, 

"Implements an Extension from the Extension Service via the ServiceExtensionProviderProps."

At least for now and we can figure out better naming in the next release.

File: core/frontend/src/extension/providers/ExtensionLoadScript.ts
Comment (kckst8) [reply core/frontend/src/extension/providers/ExtensionLoadScript.ts:17]: > @kckst8 @calebmshafer be aware that I had to comment out the original dynamic import. Otherwise, the build will not pass (specifically, the extract-api step) because of a warning "Critical Dependency: the request of a dependency is an expression". I'm still trying to figure out a way around this... I don't think we should just ignore the warning, do we?

It's likely because we are importing from a variable. I think it's safe to disable if possible. Is it coming from tsc or the linter?

Comment (johnnyd710) [reply core/frontend/src/extension/providers/ExtensionLoadScript.ts:17]: Actually I think its coming from Webpack? Is that possible? I tried ts-ignore and lint ignore comments but both did not work.

File: core/frontend/src/extension/Extension.ts
Comment (calebmshafer) [reply core/frontend/src/extension/Extension.ts]: Yea, let's go with that for now.

File: core/frontend/src/extension/providers/ExtensionServiceClient.ts
Comment (calebmshafer) [reply core/frontend/src/extension/providers/ExtensionServiceClient.ts:1]: resolving this...

File: core/frontend/src/extension/providers/ExtensionLoadScript.ts
Comment (calebmshafer) [reply core/frontend/src/extension/providers/ExtensionLoadScript.ts:17]: is this something that will prevent runtime extensions from working in general? If not, is there any future breaking change that would come from re-introducing this?

I'm fairly certain the answer is no for both. If that's the case, let's leave it out for now and merge this PR for the 3.2.0 release and file this to follow up on in 3.3.0.

Comment (johnnyd710) [reply core/frontend/src/extension/providers/ExtensionLoadScript.ts:17]: Yup, I agree. Lets leave it out for now, and it shouldn't break anything once I figure out a way to add it back without the warning.

Comment (calebmshafer) [reply core/frontend/src/extension/providers/ExtensionLoadScript.ts:17]: cool, going to resolve this and finish my review.

**Review by calebmshafer — APPROVED** [review]

File: core/frontend/src/extension/providers/ExtensionLoadScript.ts
Comment (calebmshafer) [reply core/frontend/src/extension/providers/ExtensionLoadScript.ts:17]: Can you make some issues for the various things talked about in this thread?

Comment (johnnyd710) [reply core/frontend/src/extension/providers/ExtensionLoadScript.ts:17]: Yup, I'm on it!

**Comment by pmconne** [comment]
@johnnyd710 what are we waiting on before merging this?

**Pull request merged**
